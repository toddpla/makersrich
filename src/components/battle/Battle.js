import React, { Component } from 'react';
import { connect } from 'react-redux'
import database, { firebase } from '../../firebase/firebase'
import moment from 'moment'
import selectRandom from '../../utils/selectRandom'
import { startCreditPlayer, startDebitPlayer, updatePlayer } from '../../actions/auth'
import { startSendNewsfeedMessage } from '../../actions/newsfeed'
import Timer from './Timer'
import { DEFAULT_BATTLE_BET_AMOUNT } from '../../constants'
import './rps.css'

export class Battle extends Component {

  constructor(props) {
    const openingStatements = ["Time to fight", "Choose your weapon", "What's it gonna to be Maker..."]
    super(props)
    this.state = {
      uid: props.player.uid,
      opponentUid: props.player.battle.opponentUid,
      opponentName: props.player.battle.opponentName.split(" ")[0] || '',
      // infoMessage: openingStatements[Math.floor(Math.random() * openingStatements.length)],
      infoMessage: "Choose your weapon",
      drawingStatements: ["It's === have another go", "Its a draw, go again!", "You chose the same. Have another go!"],
      waitingStatemnets: ["Weapon selected!", "Waiting for opponent!", "Here we go!"],
      winningStatements: ["Winner winer chicken dinner!", "Booyakasha - you da boss!", "YEEEEAASS! Win win win!"],
      losingStatements: ["You are a LOOOSER!", "Better luck next some Maker", "Oh dear, what have you done!"],
      betAmount: DEFAULT_BATTLE_BET_AMOUNT
    }
  }

  updateBattle = (uid, updates) => {
    database.ref(`/battles/${uid}`).update(updates)
  }

  removeBattles = () => {
    return Promise.all([
      database.ref(`/battles/${this.state.uid}`).remove(),
      database.ref(`/battles/${this.state.opponentUid}`).remove()
    ])
  }

  getOpponentWeapon = () => {
    return database.ref(`/battles/${this.state.opponentUid}/weapon`).once('value')
  }

  getPlayerWeapon = () => {
    return database.ref(`/battles/${this.state.uid}/weapon`).once('value')
  }


  finishGame = (winner) => {
    this.props.updatePlayer({battle: null})
    this.removeBattles().then(() => {
      if (winner === 'player') {
        this.props.startCreditPlayer(this.state.betAmount)
        this.props.startSendNewsfeedMessage(`beat ${this.state.opponentName} in a battle`)
      } else if (winner === 'opponent') {
        this.props.startSendNewsfeedMessage(`was beaten by ${this.state.opponentName} in a battle`)
        this.props.startDebitPlayer(this.state.betAmount)
      } else {
        this.props.startSendNewsfeedMessage(`drew with ${this.state.opponentName} in a battle`)
        this.props.startDebitPlayer(this.state.betAmount)
        // this.props.startDebitOpponent(this.state.betAmount)
      }

    })
  }

  sendChoice = (e) => {
    e.preventDefault()
    const weapon = e.target.value
    this.getOpponentWeapon().then(snap => {
      const opponentWeapon = snap.val()
      const weaponsMatrix = {'Rock': ['Scissors'], 'Paper': ['Rock'], 'Scissors': ['Paper']}
      if (Object.keys(weaponsMatrix).includes(opponentWeapon)) {
        if (weapon === opponentWeapon) {
          this.updateBattle(this.state.opponentUid, {
            infoMessage: "Its a draw, go again!",
            // infoMessage: selectRandom(this.state.drawingStatements),
            weapon: null
          })
          this.updateBattle(this.state.uid, {
            infoMessage: "Its a draw, go again!",
            // infoMessage: selectRandom(this.state.drawingStatements),
            weapon: null
          })
          return
        } else {
          const winner = (weaponsMatrix[weapon].includes(opponentWeapon)) ? 'player' : 'opponent'
          this.finishGame(winner)
        }
      } else {
        this.updateBattle(this.state.uid, {
          infoMessage: "Winner winer chicken dinner!",
          // infoMessage: selectRandom(this.state.waitingStatemnets),
          weapon
        })
      }
    })
  }

  handleTimeout = () => {
    this.getPlayerWeapon().then(snap => {
      if(snap.exists()) {
        this.finishGame('player')
      } else {
        this.finishGame()
      }
    })
  }

  render() {
    return (
      <div id="rps-battle-container">
        {this.props.player.battle ? (
          <div className='rps'>
            <h5>Battle with {this.props.player.battle.opponentName}</h5>
            <h3>{this.props.player.battle.infoMessage || this.state.infoMessage}</h3>
            <button className="rps-choice Rock" value='Rock' onClick={this.sendChoice}></button>
            <button className="rps-choice Paper" value='Paper' onClick={this.sendChoice}></button>
            <button className="rps-choice Scissors" value='Scissors' onClick={this.sendChoice}></button>
            <Timer
              end={moment(this.props.player.battle.created_at).add(5, 'seconds')}
              handleTimeout={this.handleTimeout}
            />
          </div>
        ) : (
          <h1>Game over</h1>
        )
        }
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startDebitPlayer: (cash) => dispatch(startDebitPlayer(cash)),
  startCreditPlayer: (cash) => dispatch(startCreditPlayer(cash)),
  updatePlayer: (updates) => dispatch(updatePlayer(updates)),
  startSendNewsfeedMessage: (message) => dispatch(startSendNewsfeedMessage(message))
})

const mapStateToProps = (state) => ({
  player: state.auth
})

export default connect(mapStateToProps, mapDispatchToProps)(Battle);
