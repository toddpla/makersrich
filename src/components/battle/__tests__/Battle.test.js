import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { Battle } from '../Battle'
import playersData from '../../../test/fixtures/players'
import battlesData from '../../../test/fixtures/battles'

let battle, player, startCreditPlayer, startDebitPlayer;

beforeEach(() => {
  player = playersData.fixtures[0]
  player['battle'] = battlesData.fixtures[0]
  startDebitPlayer = jest.fn()
  startCreditPlayer = jest.fn()
  battle = shallow(
    <Battle
      player={player}
      startDebitPlayer={startDebitPlayer}
      startCreditPlayer={startCreditPlayer}
    />)
})

// afterEach((done) => {
//   database.ref('/players').remove().then(() => done())
//   database.ref('/battle').remove().then(() => done())
// })

test('renders correctly', () => {
  expect(battle).toMatchSnapshot()
})

test("should win game if player choses rock and opponent chooses scissors", () => {
  console.log(battle);
})
