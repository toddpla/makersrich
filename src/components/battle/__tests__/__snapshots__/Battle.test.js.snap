// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Battle
    player={
      Object {
        "battle": Object {
          "opponentName": "James Bond",
          "opponentUid": 1,
        },
        "cash": 0,
        "displayName": "Frankie",
        "left": 0,
        "level": 1,
        "state": "online",
        "top": 0,
        "uid": "1",
      }
    }
    startCreditPlayer={[MockFunction]}
    startDebitPlayer={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="rps"
      >
        <h5>
          Battle with 
          James Bond
        </h5>
        <h3>
          Choose your weapon
        </h3>
        <Timer
          end={"1970-01-01T00:00:05.000Z"}
          handleTimeout={[Function]}
        />
        <button
          className="rps-choice Rock"
          onClick={[Function]}
          value="Rock"
        />
         
        <br />
        <button
          className="rps-choice Paper"
          onClick={[Function]}
          value="Paper"
        />
         
        <br />
        <button
          className="rps-choice Scissors"
          onClick={[Function]}
          value="Scissors"
        />
      </div>,
      "id": "rps-battle-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h5>
            Battle with 
            James Bond
          </h5>,
          <h3>
            Choose your weapon
          </h3>,
          <Timer
            end={"1970-01-01T00:00:05.000Z"}
            handleTimeout={[Function]}
          />,
          <button
            className="rps-choice Rock"
            onClick={[Function]}
            value="Rock"
          />,
          " ",
          <br />,
          <button
            className="rps-choice Paper"
            onClick={[Function]}
            value="Paper"
          />,
          " ",
          <br />,
          <button
            className="rps-choice Scissors"
            onClick={[Function]}
            value="Scissors"
          />,
        ],
        "className": "rps",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Battle with ",
              "James Bond",
            ],
          },
          "ref": null,
          "rendered": Array [
            "Battle with ",
            "James Bond",
          ],
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Choose your weapon",
          },
          "ref": null,
          "rendered": "Choose your weapon",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "end": "1970-01-01T00:00:05.000Z",
            "handleTimeout": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "rps-choice Rock",
            "onClick": [Function],
            "value": "Rock",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "rps-choice Paper",
            "onClick": [Function],
            "value": "Paper",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "rps-choice Scissors",
            "onClick": [Function],
            "value": "Scissors",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="rps"
        >
          <h5>
            Battle with 
            James Bond
          </h5>
          <h3>
            Choose your weapon
          </h3>
          <Timer
            end={"1970-01-01T00:00:05.000Z"}
            handleTimeout={[Function]}
          />
          <button
            className="rps-choice Rock"
            onClick={[Function]}
            value="Rock"
          />
           
          <br />
          <button
            className="rps-choice Paper"
            onClick={[Function]}
            value="Paper"
          />
           
          <br />
          <button
            className="rps-choice Scissors"
            onClick={[Function]}
            value="Scissors"
          />
        </div>,
        "id": "rps-battle-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h5>
              Battle with 
              James Bond
            </h5>,
            <h3>
              Choose your weapon
            </h3>,
            <Timer
              end={"1970-01-01T00:00:05.000Z"}
              handleTimeout={[Function]}
            />,
            <button
              className="rps-choice Rock"
              onClick={[Function]}
              value="Rock"
            />,
            " ",
            <br />,
            <button
              className="rps-choice Paper"
              onClick={[Function]}
              value="Paper"
            />,
            " ",
            <br />,
            <button
              className="rps-choice Scissors"
              onClick={[Function]}
              value="Scissors"
            />,
          ],
          "className": "rps",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Battle with ",
                "James Bond",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Battle with ",
              "James Bond",
            ],
            "type": "h5",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Choose your weapon",
            },
            "ref": null,
            "rendered": "Choose your weapon",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "end": "1970-01-01T00:00:05.000Z",
              "handleTimeout": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "rps-choice Rock",
              "onClick": [Function],
              "value": "Rock",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "rps-choice Paper",
              "onClick": [Function],
              "value": "Paper",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "rps-choice Scissors",
              "onClick": [Function],
              "value": "Scissors",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
